eventstore/model.go

```diff
@@ -317,24 +344,67 @@ func (t *Txn) Discard() {
 	t.discarded = true
 }
 
+type structWithID struct {
+	ID string
+}
+
 func getEntityID(t interface{}) core.EntityID {
-	v := reflect.ValueOf(t)
-	if v.Type().Kind() != reflect.Ptr {
-		v = reflect.New(reflect.TypeOf(v))
-	}
-	v = v.Elem().FieldByName(idFieldName)
-	if !v.IsValid() || v.Type() != reflect.TypeOf(core.EmptyEntityID) {
-		panic("invalid instance: doesn't have EntityID attribute")
+	// JSONMODE: assumes true. make `getEntity` a method of `Model` so maybe can access
+	// the attr, or keep it as a func and receive jsonMode as param
+	jsonMode := true
+	if jsonMode {
+		partialStruct := &structWithID{ID: "YOUDONTHAVEID"}
+		strJson := t.(*string)
+		if err := json.Unmarshal([]byte(*strJson), partialStruct); err != nil {
+			panic(err)
+		}
+		if partialStruct.ID == "YOUDONTHAVEID" {
+			panic("invalid instance: doesn't have an ID attribute")
+		}
+		// Do some extra checking that partialStruct.ID is empty, or a valid uuid, etc.
+
+		return core.EntityID(partialStruct.ID)
+
+	} else {
+
+		v := reflect.ValueOf(t)
+		if v.Type().Kind() != reflect.Ptr {
+			v = reflect.New(reflect.TypeOf(v))
+		}
+		v = v.Elem().FieldByName(idFieldName)
+		if !v.IsValid() || v.Type() != reflect.TypeOf(core.EmptyEntityID) {
+			panic("invalid instance: doesn't have EntityID attribute")
+		}
+
+		return core.EntityID(v.String())
+
 	}
-	return core.EntityID(v.String())
 }
 
 func setNewEntityID(t interface{}) core.EntityID {
-	v := reflect.ValueOf(t)
-	if v.Type().Kind() != reflect.Ptr {
-		v = reflect.New(reflect.TypeOf(v))
-	}
 	newID := core.NewEntityID()
-	v.Elem().FieldByName(idFieldName).Set(reflect.ValueOf(newID))
+	// JSONMODE: assumes true. make `getEntity` a method of `Model` so maybe can access
+	// the attr, or keep it as a func and receive jsonMode as param
+	jsonMode := true
+	if jsonMode {
+		// Ok so here we should generate a new ID and embbed it in `t`...
+		// first idea that comes to my mind. Create a json patch and apply it :)
+		jsonPatch := fmt.Sprintf(`{"ID": %q}`, newID.String())
+		strJson := t.(*string)
+		patchedValue, err := jsonpatch.MergePatch([]byte(*strJson), []byte(jsonPatch))
+		if err != nil {
+			panic(err)
+		}
+		v := reflect.ValueOf(t)
+		strPatchedValue := string(patchedValue)
+		reflectPatchedValue := reflect.ValueOf(&strPatchedValue)
+		v.Elem().Set(reflectPatchedValue.Elem())
+	} else {
+		v := reflect.ValueOf(t)
+		if v.Type().Kind() != reflect.Ptr {
+			v = reflect.New(reflect.TypeOf(v))
+		}
+		v.Elem().FieldByName(idFieldName).Set(reflect.ValueOf(newID))
+	}
```

Comment:
`setNewEntityID` which autogenerates ids...
In JSON mode, will inject the generated ID in the same `*string` instance that was called on `Create`.
leveraged a json patch to do that change